components/webext-storage/src/sync/outgoing.rs

```diff
@@ -6,94 +6,76 @@
 // managing the sync state of the local DB.
 
 use interrupt_support::Interruptee;
-use rusqlite::{Row, Transaction};
+use rusqlite::{Connection, Row, Transaction};
 use sql_support::ConnExt;
-use sync15_traits::ServerTimestamp;
+use sync15_traits::Payload;
 use sync_guid::Guid as SyncGuid;
 
 use crate::error::*;
 
-use super::ServerPayload;
+use super::Record;
 
-fn outgoing_from_row(row: &Row<'_>) -> Result<(ServerPayload, i32)> {
-    let mirror_guid = row.get::<_, Option<SyncGuid>>("mirror_guid")?;
-    let staging_guid = row.get::<_, Option<SyncGuid>>("staging_guid")?;
-    // We could assert they are identical if non-none?
-    let guid = mirror_guid
-        .or(staging_guid)
-        .unwrap_or_else(SyncGuid::random);
+fn outgoing_from_row(row: &Row<'_>) -> Result<Payload> {
+    let guid: SyncGuid = row.get("guid")?;
     let ext_id: String = row.get("ext_id")?;
     let raw_data: Option<String> = row.get("data")?;
-    let (data, deleted) = if raw_data.is_some() {
-        (raw_data, false)
-    } else {
-        (None, true)
-    };
-    let change_counter = row.get::<_, i32>("sync_change_counter")?;
-    Ok((
-        ServerPayload {
+    let payload = match raw_data {
+        Some(raw_data) => Payload::from_record(Record {
             ext_id,
             guid,
-            data,
-            deleted,
-            last_modified: ServerTimestamp(0),
-        },
-        change_counter,
-    ))
+            data: Some(raw_data),
+        })?,
+        None => Payload::new_tombstone(guid),
+    };
+    Ok(payload)
 }
 
-/// Gets info about what should be uploaded and also records metadata about these
-/// items in a temp table. Returns a vec of the payloads which
-/// should be uploaded. record_uploaded() can be called after the upload is
-/// complete and the data in the temp table will be used to update the local
-/// store.
-#[allow(dead_code)] // Kill this annotation once the bridged engine is hooked up
-pub fn get_and_record_outgoing(
-    tx: &Transaction<'_>,
-    _signal: &dyn Interruptee,
-) -> Result<Vec<ServerPayload>> {
-    // The item may not yet have a GUID (ie, it might not already be in either
-    // the mirror nor the incoming staging table.) We could probably perform
-    // some impressive sql gymnastics to handle this and arrange for the
```

Comment:
I did the SQL gymnastics only because replacing `ServerPayload` with `Payload` was kind of annoying—we have to poke at its `.data` field directly to read out the fields, or call `into_record` when we _just_ had all the fields for the record...or use `Record`, but then it's clunky for tombstones. Since the only thing we needed to make it work was the `coalesce()`, I just went for it! 😆

I also moved writing back staged incoming into the mirror here, since we merged it correctly, and waiting for `record_uploaded` means our mirror would be out of sync with what's on the server if we're interrupted mid-sync.